server/src/main/java/org/elasticsearch/search/aggregations/metrics/GlobalOrdCardinalityAggregator.java

```diff
@@ -0,0 +1,171 @@
+    private final int maxOrd;
+    private ObjectArray<BitArray> visitedOrds;
+    private final int precision;
+
+    public GlobalOrdCardinalityAggregator(
+            String name,
+            ValuesSource.Bytes.WithOrdinals valuesSource,
+            int precision,
+            int maxOrd,
+            SearchContext context,
+            Aggregator parent,
+            Map<String, Object> metadata) throws IOException {
+        super(name, context, parent, metadata);
+        this.valuesSource = valuesSource;
+        this.precision = precision;
+        this.maxOrd = maxOrd;
+        this.bigArrays = context.bigArrays();
+        this.visitedOrds = bigArrays.newObjectArray(1);
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        values = valuesSource.globalOrdinalsValues(ctx);
+        return new LeafBucketCollector() {
+            @Override
+            public void collect(int doc, long bucketOrd) throws IOException {
+                visitedOrds = bigArrays.grow(visitedOrds, bucketOrd + 1);
+                BitArray bits = visitedOrds.get(bucketOrd);
+                if (bits == null) {
+                    bits = new BitArray(maxOrd, bigArrays);
+                    visitedOrds.set(bucketOrd, bits);
+                }
+                if (values.advanceExact(doc)) {
+                    for (long ord = values.nextOrd(); ord != SortedSetDocValues.NO_MORE_ORDS; ord = values.nextOrd()) {
+                        bits.set((int) ord);
+                    }
+                }
+            }
+        };
+    }
+
+    @Override
+    protected void doPostCollection() throws IOException {
+        counts = new HyperLogLogPlusPlusSparse(precision, bigArrays, visitedOrds.size());
+        try (BitArray allVisitedOrds = new BitArray(maxOrd, bigArrays)) {
+            for (long bucket = visitedOrds.size() - 1; bucket >= 0; --bucket) {
+                final BitArray bits = visitedOrds.get(bucket);
+                if (bits != null) {
+                    allVisitedOrds.or(bits);
+                }
+            }
+
+            try (LongArray hashes = bigArrays.newLongArray(maxOrd, false)) {
+                final MurmurHash3.Hash128 hash = new MurmurHash3.Hash128();
+                for (long ord = allVisitedOrds.nextSetBit(0); ord < Long.MAX_VALUE;
+                     ord = ord + 1 < maxOrd ? allVisitedOrds.nextSetBit(ord + 1) : Long.MAX_VALUE) {
+                    final BytesRef value = values.lookupOrd(ord);
+                    MurmurHash3.hash128(value.bytes, value.offset, value.length, 0, hash);
+                    hashes.set(ord, hash.h1);
+                }
+                for (long bucket = visitedOrds.size() - 1; bucket >= 0; --bucket) {
```

Comment:
I think we cold free `allVisitedOrds` here. Do you think it'd be worth shifting the method around so `LongArray hashes` is the outer most `try`?